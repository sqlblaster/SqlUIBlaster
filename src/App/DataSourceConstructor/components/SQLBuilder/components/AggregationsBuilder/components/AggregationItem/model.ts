import { Column } from 'src/App/DataSourceConstructor/schemas';
import * as uuid from 'uuid';
import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { AggregationOperator } from '../../aggregations';

export class Aggregation {

  public id: string;
  public operator: AggregationOperator;
  public column: ColumnBranch | undefined;
  public columnFilterTypes: ColumnFilterType[];

  constructor(...inits: Partial<Aggregation>[]) {
    Object.assign(this, ...inits);

    if (!this.id) {
      this.id = uuid();
    }

    this.columnFilterTypes = Aggregation.getExpectedColumnsTypeFrom(
      this.operator
    );
    !this.columnFilterTypes.length && (this.column = undefined);
  }

  public static hasNoColumn(operator: AggregationOperator): boolean {
    return (
      operator === 'Count of rows' || operator === 'Cumulative count of rows'
    );
  }

  public static getExpectedColumnsTypeFrom(
    operator: AggregationOperator
  ): Aggregation['columnFilterTypes'] {
    if (operator === 'Number of distinct values of') {
      return ['all'];
    } else if (!Aggregation.hasNoColumn(operator)) {
      return ['number'];
    } else {
      return [];
    }
  }
}

export type ColumnFilterType = Column['type'] | 'all' | undefined;
