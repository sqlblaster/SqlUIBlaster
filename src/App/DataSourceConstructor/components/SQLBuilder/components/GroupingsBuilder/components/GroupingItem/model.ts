import uuid from 'uuid';
import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';

export class Grouping {
  public id: string;
  // TODO: rename column to columnBranch
  public column: ColumnBranch;

  constructor(...inits: Partial<Grouping>[]) {
    Object.assign(this, ...inits);

    if (!this.id) {
      this.id = uuid();
    }
  }
}
