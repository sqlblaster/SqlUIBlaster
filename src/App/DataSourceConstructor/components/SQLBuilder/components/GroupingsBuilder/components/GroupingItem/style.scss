.grouping-item {
  display: flex;
  align-items: center;
  border: 2px solid transparent;
  border-radius: 3px;
  transition: 0.3s;
  margin: 5px 0px;
}
