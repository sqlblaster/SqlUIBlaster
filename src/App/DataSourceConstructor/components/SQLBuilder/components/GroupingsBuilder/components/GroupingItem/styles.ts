import { StyleRulesCallback } from '@material-ui/core';

export type GroupingItemClassKeys = 'root' | 'column-view';

export const groupingItemStyles: StyleRulesCallback<
  GroupingItemClassKeys
> = (theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    padding: 2,
    borderRadius: '3px',
    transition: '0.3s',
    margin: '5px 0px'
  },
  'column-view': {
    cursor: 'pointer',
    '&:hover': {
      opacity: 0.8
    },
    transition: '0.3s'
  }
});
