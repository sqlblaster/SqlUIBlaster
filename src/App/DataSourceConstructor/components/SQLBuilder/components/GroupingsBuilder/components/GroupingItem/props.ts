import { ClassesProp } from 'src/App/utils/classes-prop';
import { actionCreators, PickActionCreators } from 'src/store/action-creators';
import { PopoverManagerProps } from '../../../common/WithPopoverManagement';
import { Grouping } from './model';
import { GroupingItemClassKeys } from './styles';

export interface GroupingItemOwnProps {
  grouping: Grouping;
}

export type GroupingItemDispatchProps = PickActionCreators<
  'removeGrouping' | 'updateGrouping'
>;

export const GIMapDispatchToProps: GroupingItemDispatchProps = {
  removeGrouping: actionCreators.removeGrouping,
  updateGrouping: actionCreators.updateGrouping
};

export type GroupingItemProps = GroupingItemOwnProps &
  GroupingItemDispatchProps;

export type GroupingItemViewProps = {
  handleGroupingRemoval: (id: Grouping['id']) => () => void;
} & GroupingItemOwnProps &
  PopoverManagerProps &
  PickActionCreators<'updateGrouping'> &
  ClassesProp<GroupingItemClassKeys>;
