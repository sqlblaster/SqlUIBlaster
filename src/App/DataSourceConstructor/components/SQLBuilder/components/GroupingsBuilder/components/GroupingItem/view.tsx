import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { RemoveButton } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnView } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/SequentialColumnView';
import { MutateGroupingPopover } from '../MutateGroupingPopover';
import { GroupingItemViewProps } from './props';
import { groupingItemStyles } from './styles';

export const GroupingItem = withStyles(groupingItemStyles)(
  class extends React.Component<GroupingItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: GroupingItemViewProps) {
      super(props);

      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        grouping,
        updateGrouping,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleGroupingRemoval
      } = this.props;

      const { id, column } = grouping;

      return (
        <>
          <div className={classes.root}>
            <div
              ref={this.popoverLauncher}
              onClick={handleOpen}
              className={classes['column-view']}
              data-testid='column-view'
            >
              <SequentialColumnView {...column} />
            </div>
            <RemoveButton
              onClick={handleGroupingRemoval(id)}
              data-testid='remove'
            />
          </div>
          { isOpen && (
            <MutateGroupingPopover
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateGrouping}
              onClose={handleClose}
              grouping={grouping}
            />
          ) }
        </>
      );
    }
  }
);
