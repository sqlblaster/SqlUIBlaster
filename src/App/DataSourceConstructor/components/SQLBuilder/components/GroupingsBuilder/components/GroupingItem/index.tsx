import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import {
  GIMapDispatchToProps,
  GroupingItemDispatchProps,
  GroupingItemOwnProps,
  GroupingItemProps,
  GroupingItemViewProps as ViewProps
} from './props';
import { GroupingItem as View } from './view';

export const GroupingItem = connect<
  {},
  GroupingItemDispatchProps,
  GroupingItemOwnProps,
  State
>(
  null,
  GIMapDispatchToProps
)(
  class extends React.Component<GroupingItemProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public handleGroupingRemoval: ViewProps['handleGroupingRemoval'] = (id) => () =>
      this.props.removeGrouping(id)

    public render() {
      return (
        <GroupingItem.ViewWithPopoverManagement
          handleGroupingRemoval={this.handleGroupingRemoval}
          {...this.props}
        />
      );
    }
  }
);
