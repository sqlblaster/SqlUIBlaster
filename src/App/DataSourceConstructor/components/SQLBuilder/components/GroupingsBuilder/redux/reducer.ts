import { Reducer, ReducersMapObject } from 'redux';
import { Grouping } from '../components/GroupingItem/model';
import {
  AddGroupingAction,
  GroupingActions,
  RemoveGroupingAction,
  UpdateGroupingAction
} from './action';
import { GroupingsBuilderState } from './state';

export const addGrouping: Reducer<Grouping[], AddGroupingAction> = (
  state = [],
  { grouping }
) => state.concat(new Grouping({ ...grouping }));

export const updateGrouping: Reducer<Grouping[], UpdateGroupingAction> = (
  state = [],
  { grouping }
) =>
  state.map((a) => {
    if (a.id !== grouping.id) {
      return a;
    }

    return new Grouping(a, grouping);
  });

export const removeGrouping: Reducer<Grouping[], RemoveGroupingAction> = (
  state = [],
  { id }
) => state.filter((grouping) => grouping.id !== id);

export const groupings: Reducer<
  Grouping[],
  GroupingActions
> = (state = [], action) => {
  switch (action.type) {
    case 'ADD_GROUPING':
      return addGrouping(state, action);
    case 'UPDATE_GROUPING':
      return updateGrouping(state, action);
    case 'REMOVE_GROUPING':
      return removeGrouping(state, action);
    case 'CLEAR_GROUPINGS':
      return [];
    default:
      return state;
  }
};

export const GroupingsBuilderReducers: ReducersMapObject<
  GroupingsBuilderState
> = {
  groupings
};
