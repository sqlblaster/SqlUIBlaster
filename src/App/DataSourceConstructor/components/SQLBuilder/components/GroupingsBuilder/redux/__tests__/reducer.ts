import { testInitialAndCurrentState } from 'src/components/query-builder/App/DataSourceConstructor/test-utils/store/reducers';
import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { Grouping } from '../../components/GroupingItem/model';
import {
  addGrouping,
  clearGroupings,
  removeGrouping,
  updateGrouping
} from '../action';
import { groupings as reducer } from '../reducer';

describe('groupings reducer', () => {
  testInitialAndCurrentState(reducer, [], [new Grouping(), new Grouping()]);

  it('should add new grouping on ADD_GROUPING action', () => {
    expect(
      reducer(
        [],
        addGrouping(
          new Grouping({ id: '1' })
        )
      )
    ).toMatchSnapshot();

    expect(
      reducer(
        [
          new Grouping({ id: '1' }),
          new Grouping({ id: '2' })
        ],
        addGrouping(
          new Grouping({ id: '3' })
        )
      )
    ).toMatchSnapshot();
  });

  it('should update grouping on UPDATE_GROUPING action', () => {
    expect(
      reducer([], updateGrouping(new Grouping({ id: '4' })))
    ).toMatchSnapshot();

    expect(
      reducer(
        [new Grouping({ id: '3' }), new Grouping({ id: '4' })],
        updateGrouping(
          new Grouping({
            id: '4',
            column: new ColumnBranch({ columnName: 'ID' })
          })
        )
      )
    ).toMatchSnapshot();

    expect(
      reducer(
        [new Grouping({ id: '3' }), new Grouping({ id: '5' })],
        updateGrouping(
          new Grouping({
            id: '4',
            column: new ColumnBranch({ columnName: 'ID' })
          })
        )
      )
    ).toMatchSnapshot();
  });

  it('should remove grouping by id on REMOVE_GROUPING action', () => {
    expect(reducer([], removeGrouping('2'))).toMatchSnapshot();

    expect(
      reducer(
        [new Grouping({ id: '1' }), new Grouping({ id: '2' })],
        removeGrouping('4')
      )
    ).toMatchSnapshot();

    expect(
      reducer(
        [new Grouping({ id: '4' }), new Grouping({ id: '2' })],
        removeGrouping('4')
      )
    ).toMatchSnapshot();
  });

  it('should clear all groupings on CLEAR_GROUPINGS action', () => {
    expect(reducer([], clearGroupings())).toMatchSnapshot();

    expect(
      reducer([new Grouping(), new Grouping()], clearGroupings())
    ).toMatchSnapshot();
  });
});
