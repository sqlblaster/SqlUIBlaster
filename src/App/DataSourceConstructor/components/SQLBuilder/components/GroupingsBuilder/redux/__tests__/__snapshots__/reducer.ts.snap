// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupings reducer should add new grouping on ADD_GROUPING action 1`] = `
Array [
  Grouping {
    "id": "1",
  },
]
`;

exports[`groupings reducer should add new grouping on ADD_GROUPING action 2`] = `
Array [
  Grouping {
    "id": "1",
  },
  Grouping {
    "id": "2",
  },
  Grouping {
    "id": "3",
  },
]
`;

exports[`groupings reducer should clear all groupings on CLEAR_GROUPINGS action 1`] = `Array []`;

exports[`groupings reducer should clear all groupings on CLEAR_GROUPINGS action 2`] = `Array []`;

exports[`groupings reducer should remove grouping by id on REMOVE_GROUPING action 1`] = `Array []`;

exports[`groupings reducer should remove grouping by id on REMOVE_GROUPING action 2`] = `
Array [
  Grouping {
    "id": "1",
  },
  Grouping {
    "id": "2",
  },
]
`;

exports[`groupings reducer should remove grouping by id on REMOVE_GROUPING action 3`] = `
Array [
  Grouping {
    "id": "2",
  },
]
`;

exports[`groupings reducer should update grouping on UPDATE_GROUPING action 1`] = `Array []`;

exports[`groupings reducer should update grouping on UPDATE_GROUPING action 2`] = `
Array [
  Grouping {
    "id": "3",
  },
  Grouping {
    "column": Object {
      "_foreignTableName": undefined,
      "_lastColumn": undefined,
      "columnName": "ID",
      "foreignColumn": undefined,
      "foreignModelName": undefined,
      "foreignTableName": undefined,
      "modelName": "",
      "tableName": "",
      "type": undefined,
    },
    "id": "4",
  },
]
`;

exports[`groupings reducer should update grouping on UPDATE_GROUPING action 3`] = `
Array [
  Grouping {
    "id": "3",
  },
  Grouping {
    "id": "5",
  },
]
`;
